netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -198,6 +313,243 @@ _net_ebpf_sock_addr_update_store_entries()
+    InsertTailList(&_net_ebpf_ext_redirect_handle_list, &entry->list_entry);
+    ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock, old_irql);
+
+Exit:
+    return status;
+}
+
+static void
+_net_ebpf_ext_sock_addr_delete_redirect_handle(uint64_t filter_id)
+{
+    KIRQL old_irql;
+
+    old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
+
+    LIST_ENTRY* list_entry = _net_ebpf_ext_redirect_handle_list.Flink;
+    while (list_entry != &_net_ebpf_ext_redirect_handle_list) {
+        net_ebpf_extension_redirect_handle_entry_t* entry =
+            CONTAINING_RECORD(list_entry, net_ebpf_extension_redirect_handle_entry_t, list_entry);
+        if (entry->filter_id == filter_id) {
+            RemoveEntryList(list_entry);
+            ExFreePool(entry);
+            break;
+        }
+        list_entry = list_entry->Flink;
+    }
+    ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock, old_irql);
+}
+
+static NTSTATUS
+_net_ebpf_ext_sock_addr_get_redirect_handle(uint64_t filter_id, _Out_ HANDLE* redirect_handle)
+{
+    NTSTATUS status = STATUS_NOT_FOUND;
+    KIRQL old_irql;
+
+    old_irql = ExAcquireSpinLockShared(&_net_ebpf_ext_sock_addr_lock);
+
+    LIST_ENTRY* list_entry = _net_ebpf_ext_redirect_handle_list.Flink;
+    while (list_entry != &_net_ebpf_ext_redirect_handle_list) {
+        net_ebpf_extension_redirect_handle_entry_t* entry =
+            CONTAINING_RECORD(list_entry, net_ebpf_extension_redirect_handle_entry_t, list_entry);
+        if (entry->filter_id == filter_id) {
+            *redirect_handle = entry->redirect_handle;
+            status = STATUS_SUCCESS;
+            break;
+        }
+        list_entry = list_entry->Flink;
+    }
+
+    ExReleaseSpinLockShared(&_net_ebpf_ext_sock_addr_lock, old_irql);
+
+    return status;
+}
+
+static void
+_net_ebpf_ext_connection_context_initialize_value(
+    bool redirected, uint32_t verdict, _Out_ net_ebpf_extension_connection_context_t* connection_context)
+{
+    connection_context->value.redirected = redirected;
+    connection_context->value.verdict = (uint8_t)verdict;
+    connection_context->timestamp = CONVERT_100NS_UNITS_TO_MS(KeQueryInterruptTime());
+}
+
+/**
+ * @brief Compare the destination address in the two provided bpf_sock_addr_t structs.
+ *
+ * @param[in] addr1 Pointer to first sock_addr struct to compare.
+ * @param[in] addr2 Pointer to second sock_addr struct to compare.
+
+ * @return true, if the addresses are same, false otherwise.
+ */
+static inline bool
+_net_ebpf_ext_compare_destination_address(_In_ const bpf_sock_addr_t* addr1, _In_ const bpf_sock_addr_t* addr2)
+{
+    ASSERT(addr1->family == addr2->family);
+    if (addr1->family != addr2->family) {
+        return false;
+    }
+
+    return INET_ADDR_EQUAL((ADDRESS_FAMILY)addr1->family, &addr1->user_ip4, &addr2->user_ip4);
+}
+
+static void
+_net_ebpf_ext_connection_context_initialize_key(
+    _Inout_ net_ebpf_ext_connection_context_key_t* context_key,
```

Comment:
this is not inout - if you are only using this to get the length of ip_address, you could just check the address family. This should be a out parameter and the last one in the parameter list.

Reply:
Updated